File: frontend/src/types/location.ts
import type { ProductListItem } from '@/types/product'

export type LocationType =
  | 'tea_coffee_farm'
  | 'fruit_garden'
  | 'flower_garden'
  | 'dairy_farm'
  | 'vegetable_farm'
  | 'craft_village'
  | 'farmstay'
  | 'other'

export type LocationSort = 'name' | '-name' | 'newest' | 'rating'

export interface LocationTypeOption {
  value: LocationType
  label: string
}

export interface LocationFilterOptions {
  types: LocationTypeOption[]
  districts: string[]
}

export interface LocationListItem {
  id: number
  name: string
  slug: string
  type: LocationType
  type_label: string
  district: string
  address: string
  latitude: number
  longitude: number
  opening_hours: string | null
  ticket_price: number | null
  services: string[]
  description: string | null
  rating_avg: number
  primary_image_url: string | null
}

export interface LocationListResponse {
  items: LocationListItem[]
  page: number
  page_size: number
  total: number
}

export interface LocationFilters {
  page?: number
  page_size?: number
  search?: string
  type?: LocationType
  district?: string
  sort?: LocationSort
}

export interface LocationImage {
  id: number
  image_url: string
  is_primary: boolean
  sort_order: number
}

export interface LocationSubject {
  id: number
  name: string
  district: string
}

export interface LocationDetail extends LocationListItem {
  contact_phone: string | null
  website: string | null
  source_url: string | null
  views: number
  images: LocationImage[]
  subject: LocationSubject | null
  products: ProductListItem[]
  updated_at: string
}

export interface MapFeatureProperties {
  id: number
  slug: string
  name: string
  type: LocationType
  type_label: string
  district: string
  address: string
  opening_hours: string | null
  ticket_price: number | null
  rating_avg: number
  primary_image_url: string | null
}

export interface MapFeature {
  type: 'Feature'
  geometry: { type: 'Point'; coordinates: [number, number] }
  properties: MapFeatureProperties
}

export interface MapFeatureCollection {
  type: 'FeatureCollection'
  features: MapFeature[]
}

export interface MapFilters {
  search?: string
  type?: LocationType
  district?: string
}

export interface Coordinate {
  latitude: number
  longitude: number
}

export interface NearbyLocation extends LocationListItem {
  distance_m: number
}

export interface NearbyLocationResponse {
  origin: Coordinate
  radius_m: number
  items: NearbyLocation[]
}

export interface RouteResponse {
  provider: 'OSRM'
  profile: 'driving'
  origin: Coordinate
  destination: { slug: string; name: string; latitude: number; longitude: number }
  distance_m: number
  duration_s: number
  geometry: { type: 'LineString'; coordinates: Array<[number, number]> }
}

File: backend/app/schemas/location.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

from app.schemas.product import ProductListItem


class LocationTypeOption(BaseModel):
    value: str
    label: str


class LocationFilterOptions(BaseModel):
    types: list[LocationTypeOption]
    districts: list[str]


class LocationListItem(BaseModel):
    id: int
    name: str
    slug: str
    type: str
    type_label: str
    district: str
    address: str
    latitude: float
    longitude: float
    opening_hours: str | None
    ticket_price: float | None
    services: list[str]
    description: str | None
    rating_avg: float
    primary_image_url: str | None


class LocationListResponse(BaseModel):
    items: list[LocationListItem]
    page: int
    page_size: int
    total: int


class LocationImageRead(BaseModel):
    id: int
    image_url: str
    is_primary: bool
    sort_order: int


class LocationSubjectRead(BaseModel):
    id: int
    name: str
    district: str


class LocationDetail(LocationListItem):
    contact_phone: str | None
    website: str | None
    source_url: str | None
    views: int
    images: list[LocationImageRead]
    subject: LocationSubjectRead | None
    products: list[ProductListItem]
    updated_at: datetime


class PointGeometryRead(BaseModel):
    type: Literal["Point"] = "Point"
    coordinates: tuple[float, float]


class MapFeatureProperties(BaseModel):
    id: int
    slug: str
    name: str
    type: str
    type_label: str
    district: str
    address: str
    opening_hours: str | None
    ticket_price: float | None
    rating_avg: float
    primary_image_url: str | None


class MapFeature(BaseModel):
    type: Literal["Feature"] = "Feature"
    geometry: PointGeometryRead
    properties: MapFeatureProperties


class MapFeatureCollection(BaseModel):
    type: Literal["FeatureCollection"] = "FeatureCollection"
    features: list[MapFeature]


class Coordinate(BaseModel):
    latitude: float
    longitude: float


class NearbyLocation(LocationListItem):
    distance_m: float


class NearbyLocationResponse(BaseModel):
    origin: Coordinate
    radius_m: float
    items: list[NearbyLocation]


class LineStringGeometryRead(BaseModel):
    type: Literal["LineString"] = "LineString"
    coordinates: list[tuple[float, float]]


class RouteDestination(BaseModel):
    slug: str
    name: str
    latitude: float
    longitude: float


class RouteResponse(BaseModel):
    provider: Literal["OSRM"] = "OSRM"
    profile: Literal["driving"] = "driving"
    origin: Coordinate
    destination: RouteDestination
    distance_m: float
    duration_s: float
    geometry: LineStringGeometryRead
